core/src/main/java/org/elasticsearch/common/xcontent/XContentParserUtils.java

```diff
@@ -121,23 +121,26 @@ public static Object parseStoredFieldsValue(XContentParser parser) throws IOExce
      * @param parser      the current {@link XContentParser}
      * @param delimiter   the delimiter to use to splits the field's name
      * @param objectClass the object class of the object to parse
+     * @param objects     a list of objects the parsed one will be added to
      * @param <T>         the type of the object to parse
      * @return the parsed object
      * @throws IOException if anything went wrong during parsing or if the type or name cannot be derived
      *                     from the field's name
      */
-    public static <T> T parseTypedKeysObject(XContentParser parser, String delimiter, Class<T> objectClass) throws IOException {
+    public static <T> void parseTypedKeysObject(XContentParser parser, String delimiter, Class<T> objectClass, List<T> objects)
```

Comment:
When do we need to only use the first element in the list? I think that happens only in tests. In which case I would not add a second method here. Let's settle on one signature and go for it. I think Consumer may make things a bit better for tests already